write-fonts/src/pens.rs

```diff
@@ -0,0 +1,369 @@
+
+    Ok(())
+}
+
+impl<'a, T: Pen> ReverseContourPen<'a, T> {
+    pub fn new(inner_pen: &'a mut T) -> ReverseContourPen<T> {
+        ReverseContourPen {
+            inner_pen,
+            pending: Vec::new(),
+        }
+    }
+
+    /// Flush buffer into inner, reversing the winding order.
+    ///
+    /// If start == end, as is typically in a [move, ..., close] structure, start point is unchanged.
+    ///
+    /// Requires an explicit call to afford the client the opportunity to receive errors.
+    pub fn flush(&mut self) -> Result<(), ContourReversalError> {
+        // Process subpath by subpath
+        let mut subpath_starts: Vec<usize> = self
+            .pending
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, cmd)| {
+                match cmd {
+                    // Move starts a new subpath
+                    // However, since we will split on idx ignore 0
+                    PenCommand::MoveTo { .. } => Some(idx),
+                    _ => None,
+                }
+            })
+            .collect();
+        subpath_starts.push(self.pending.len());
+        for win in subpath_starts.windows(2) {
+            flush_subpath(&self.pending[win[0]..win[1]], self.inner_pen)?;
+        }
+
+        self.pending.clear();
+
+        Ok(())
+    }
+}
+
+impl<'a, T: Pen> Pen for ReverseContourPen<'a, T> {
+    fn move_to(&mut self, x: f32, y: f32) {
+        self.pending.push(PenCommand::MoveTo { x, y });
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        self.pending.push(PenCommand::LineTo { x, y });
+    }
+
+    fn quad_to(&mut self, cx0: f32, cy0: f32, x: f32, y: f32) {
+        self.pending.push(PenCommand::QuadTo { cx0, cy0, x, y });
+    }
+
+    fn curve_to(&mut self, cx0: f32, cy0: f32, cx1: f32, cy1: f32, x: f32, y: f32) {
+        self.pending.push(PenCommand::CurveTo {
+            cx0,
+            cy0,
+            cx1,
+            cy1,
+            x,
+            y,
+        });
+    }
+
+    fn close(&mut self) {
+        self.pending.push(PenCommand::Close);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use font_types::Pen;
+    use kurbo::Affine;
+
+    use super::{BezPathPen, ReverseContourPen, TransformPen};
+
+    fn draw_open_test_shape(pen: &mut impl Pen) {
+        pen.move_to(10.0, 10.0);
+        pen.quad_to(40.0, 40.0, 60.0, 10.0);
+        pen.line_to(100.0, 10.0);
+        pen.curve_to(125.0, 10.0, 150.0, 50.0, 125.0, 60.0)
+    }
+
+    fn draw_closed_triangle(pen: &mut impl Pen) {
+        pen.move_to(100.0, 100.0);
+        pen.line_to(150.0, 200.0);
+        pen.line_to(50.0, 200.0);
+        pen.close();
+    }
+
+    fn draw_closed_test_shape(pen: &mut impl Pen) {
+        pen.move_to(125.0, 100.0);
+        pen.quad_to(200.0, 150.0, 175.0, 300.0);
+        pen.curve_to(150.0, 150.0, 50.0, 150.0, 25.0, 300.0);
+        pen.quad_to(0.0, 150.0, 75.0, 100.0);
+        pen.line_to(100.0, 50.0);
+        pen.close();
```

Comment:
I think I found an issue. All your closed test shapes end with a line_to so this does not show up in your tests.
Basically, whenever the last pen command before the close is _not_ a line_to (e.g. a curve_to or quad_to or move_to even), then after reversing you end up with an _additional_ line segment that overlaps the move.
E.g. given this path (notice it ends with a curve_to before the closing 'Z'):

"M100,100 L150,200 L50,200 C75,200 75,100 100,100 Z" 

If I reverse it using your current code, I get this (I added commas for legibility):

"M100,100 L100,100 C75,100 75,200 50,200 L150,200 Z"

Note that first 'L' segment following the 'M' is a duplicate and was not in the original closed path.

Compare this with the output from fontTools ReverseContourPen:

```python
from fontTools.pens.recordingPen import RecordingPen
from fontTools.pens.reverseContourPen import ReverseContourPen
rec = RecordingPen()
rev = ReverseContourPen(rec)
rev.moveTo((100.0, 100.0))
rev.lineTo((150.0, 200.0))
rev.lineTo((50.0, 200.0))
rev.curveTo((75.0, 200.0), (75.0, 100.0), (100.0, 100.0))
rev.closePath()
print(rec.value)
"""
[('moveTo', ((100.0, 100.0),)),
 ('curveTo', ((75.0, 100.0), (75.0, 200.0), (50.0, 200.0))),
 ('lineTo', ((150.0, 200.0),)),
 ('closePath', ())]
"""
```

Reply:
All this talk about implied closing line lead me to discover what I believe is a bug in fonttools' implementation as well LOL

https://github.com/fonttools/fonttools/pull/2995

The ReverseContourPen's `outputImpliedClosingLine` parameter is a very recent addition by behdad, we haven't released it yet (a parameter with that name previously existed only in the PointToSegmentPen). It turns out it has a very similar bug whereby a duplicate lineTo, overlapping moveTo, gets inserted after reversing a closed path.
The bug is not present in the default case, when outputImpliedClosingLine=False.

Reply:
> I added commas for legibility

@dfrg are we allowed to change the to_svg format for Kurbo? - it is quite illegible right now

Reply:
Forgive my density but I don't quite follow.

```
E.g. given this path (notice it ends with a curve_to before the closing 'Z'):

"M100,100 L150,200 L50,200 C75,200 75,100 100,100 Z" => MLLCZ

If I reverse it using your current code, I get this (I added commas for legibility):

"M100,100 L100,100 C75,100 75,200 50,200 L150,200 Z" => MLCLZ
```

So before and after we have 1 M, 2 L's, 1 C, and a Z. In the original the Z is pointless - we already ended the C at 100,100 - so when reversed the first command is pointless. I feel like I'm missing something.

Reply:
I don’t think kurbo makes any particular guarantees about the svg output. I assume you’d prefer a format like the comment above?

Reply:
> I assume you’d prefer a format like the comment above?

Yes. The , helps spot which values are a point and the space *before* the letter makes it *far* easier to spot the commands.